File: CPP/No446.cc
/**
 * Created by Xiaozhong on 2020/9/23.
 * Copyright (c) 2020/9/23 Xiaozhong. All rights reserved.
 */
#include "vector"
#include "iostream"
#include "unordered_map"

typedef long long ll;
using namespace std;

typedef long long ll;
class Solution {
public:
    int numberOfArithmeticSlices(vector<int> &A) {
        int n = A.size();
        ll ans = 0;
        vector<unordered_map<ll, int>> cnt(n);
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                ll delta = (ll) A[i] - (ll) A[j];
                int sum = 0;
                if (cnt[j].find(delta) != cnt[j].end()) {
                    sum += cnt[j][delta];
                }
                cnt[i][delta] += sum + 1;
                ans += sum;
            }
        }
        return (int) ans;
    }
};

File: Java/src/No446.java
/*
 * Copyright (c) 8/11/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.HashMap;
import java.util.Map;

public class No446 {
    public int numberOfArithmeticSlices(int[] nums) {
        int ans = 0;
        int n = nums.length;
        Map<Long, Integer>[] mapper = new Map[n];
        for (int i = 0; i < n; ++i) mapper[i] = new HashMap<>();
        for (int j = 0; j < n; ++j) {
            for (int i = 0; i < j; ++i) {
                long d = 1L * nums[j] - nums[i];
                int cnt = mapper[i].getOrDefault(d, 0);
                ans += cnt;
                mapper[i].put(d, mapper[j].getOrDefault(d, 0) + cnt + 1);
            }
        }
        return ans;
    }
}
